### ISSUE TITLE
Tab completion behavior when prefix matches

### ISSUE THREAD

**Issue by iamevn** [fish-shell/fish-shell#7803]
> fish versions: 3.0.2 on windows (WSL (Ubuntu)) and 3.1.2 on Debian (both tried out with clean environments)

I was tabbing through `apt` completions looking for packages where I was surprised to not see both foo* and libfoo* in completions for packages when running a completion from a command like `> apt install foo`

Current behavior: (tab completion triggered when cursor in the position `#tab` is in. shown completions listed on next line.)
```fish
> function f
    echo $argv
  end
> set -l args 123456 123abc abc abcdef
> complete -c f -f -a "$args"
> f #tab
123abc  123456  abc  abcdef
> f bc#tab
123abc  abc  abcdef
> f abc#tab
abc abcdef
```
In the last case here, it looks like the `123abc` completion was filtered out because `abc` matched the prefix of other completions. I feel like I'd rather have `123abc` available as a completion here but maybe ordered after prefix matches.

I'd like some confirmation on how tab completions work and am wondering whether changing this to include the other non-prefix completions would make sense/be desirable. (Would be happy to contribute if this sounds like a desirable thing.)

**Comment by faho** [comment]
Yes, the behavior is to show infix matches until we get a prefix match.

The issue with keeping infix (and especially fuzzy - "abc" would also match "a_bear_cool.mp3") matches around is that it increases the number of matches quite significantly and makes it hard to find the full prefix match.

The issue with not keeping them is that, if you're actually looking for infix matches (like you're typing "abc" to find "123abc") it might not show up because you have an accidental prefix match - like here.

There is a tradeoff here, sure. I'm not sure we're on the right side of it but I'm also wary of changing this now. Ideally there would be some better behavior we could find.

**Comment by iamevn** [comment]
I don't think I mind having too many completions if they're sorted by match rank and then alphabetically within those ranks. That way the most likely matches are still at the top. The only issue I see there is that shift+tab now goes to the bottom of a lot longer list.

Maybe a keybind that adds one level of worse matches to the results each time you press it?

Or maybe this is actually a documentation bug because I was surprised that completions I wrote weren't showing up and clarifying what fish does with the arguments from `complete` would help clear it up.

**Comment by devjgm** [comment]
fwiw, I was also just bitten/surprised by this. My tool was returning candidates that weren't showing up, because one of them was a prefix match that was immediately accepted. The behavior I wanted/expected was that fish would show all the completion matches that were returned.

**Comment by tv42** [comment]
I added a `complete-prefix` command, the change was very simple and bind tab `complete-prefix` seems to give me the behavior I want. I have `complete` bound to shift-tab. Try it out: https://github.com/fish-shell/fish-shell/pull/12381

Cross-linking:
- https://github.com/fish-shell/fish-shell/discussions/11670
- https://github.com/fish-shell/fish-shell/issues/7297
